### PR TITLE
[Executor] Merge sequential & parallel execution flow

### PR DESCRIPTION
**Pull request by gelash**
De-spagettify the aptos-vm execution flow 
- remove unused status (only used in harness tests, fix the tests to have the same behavior as in prod).
- merge sequential with parallel executor crate, re-using code but keep the same algorithm for redundancy / testing & fallback (i.e. Block-STM could also run sequential with one thread with probably not too much overhead). 

Benefits:
- Cleaner flow
- Can hook to the same executor regardless of whether we are executing sequential or parallel.
- Can use the powerful testing framework we have for parallel executor for sequential execution.
- Can re-use the executor rayon threadpool for block signature checks.

EDIT: These are coming in a different diff.
Additional improvements to tests:
    - Test Over / Underflows based on storage values (close to 0 or close to u128::MAX)
    - test different number of cores
    - Refactors: ModulePath enum, isolate BaselineState for generating baseline with different configurations (i.e. aggregator values materialized or not), check delta sequence, etc.
    - Make sure StorageError precedence over DeltaApplicationFailure in speculative executions when aggregator is deleted but deltas on top also fail. Not relevant for current use-case (no under/over flows and deletes), also also the proper algorithm for handling general case may not have the issue anyway.
After the testing PR, @perryjrandall we should start running on different platforms too.

### CODE REVIEW

**Comment by dariorussi** [comment]
just a quick comment as I am reviewing this, and it will be slow for me as it touches a lot of core stuff I have not seen before. Though a wonderful exercise, thanks!
I am not sure when we cut the branch (if we did not already) and all of the main-net implications, but this is cutting a bunch of core stuff that I can tell and we want it in definitely after we are done with main-net and all.
Can you share your thoughts?

**Comment by gelash** [comment]
@dariorussi - yes, if we like it, we should roll this in after main-net. Should be nothing breaking, more tests, common flow and should facilitate lots of future improvements (e.g. even just the virtue of having a single executor object for callbacks or what not).

Will def fix all the linter and related errors. And (TODO to self): will also experiment with Dashset/Dashmap configuration settings (num shards mainly), we use defaults and maybe there is some perf to squeeze.

File: aptos-move/e2e-tests/src/executor.rs
Comment (zekun000) [inline aptos-move/e2e-tests/src/executor.rs:383]: please keep the usize::min

File: aptos-move/aptos-vm/src/parallel_executor/vm_wrapper.rs
Comment (zekun000) [inline aptos-move/aptos-vm/src/parallel_executor/vm_wrapper.rs:105]: nit: probably can squeeze this in the view to be consistent

File: aptos-move/aptos-vm/src/data_cache.rs
Comment (zekun000) [inline aptos-move/aptos-vm/src/data_cache.rs:50]: there's existing type for this -> std::borrow::Cow

File: aptos-move/aptos-vm/src/parallel_executor/mod.rs
Comment (zekun000) [inline aptos-move/aptos-vm/src/parallel_executor/mod.rs:118]: there's too many indentation here, I'd also try to extract the delta handling out from this block

**Review by zekun000 — COMMENTED** [review]
early comments, need to do another pass

File: aptos-move/aptos-vm/src/data_cache.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/data_cache.rs:50]: Sweet!

File: aptos-move/aptos-vm/src/parallel_executor/mod.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/parallel_executor/mod.rs:118]: refactored the whole thing

File: aptos-move/aptos-vm/src/parallel_executor/vm_wrapper.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/parallel_executor/vm_wrapper.rs:105]: as per our discussion: for a future refactoring, we can take some view related things from aptos vm and then we can include there and in parallel executor too, and we can unify these interfaces.

File: aptos-move/e2e-tests/src/executor.rs
Comment (gelash) [reply aptos-move/e2e-tests/src/executor.rs:383]: damn, thanks for the catch. no idea what happened, maybe I messed up the rebase. I remember this is the optimization for tests.
Will also do the other check later than if block is small use sequential.

**Review by dariorussi — COMMENTED** [review]
I think you are splitting this up in multiple PRs which is a very good idea, so love it and maybe you want to mark it somehow.
Just commenting given I have a small comment I think you may enjoy.

**Comment by gelash** [comment]
> I think you are splitting this up in multiple PRs which is a very good idea, so love it and maybe you want to mark it somehow. Just commenting given I have a small comment I think you may enjoy.

Will do ASAP. To document here, the plan is to do proptest changes in a separate diff. 

I will make sure to stack the proptest PR on top though and not land the first PR without ensuring the second one (with new and stronger tests) passes.

File: aptos-move/aptos-vm/src/parallel_executor/mod.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/parallel_executor/mod.rs:170]: Perfect! This bothered me too and now that you said it, I will go and change the name parallel executor :-) Maybe we should call it block_executor. and it knows whether to execute sequentially, parallel, etc.

Moreover, I will move the post processing / delta resolution (and the Module intersection fallback hack) there as well. I believe before I had some reason why I couldn't pull it off, but I believe now it should be possible. This should reduce the duct tape logic in these connectors between aptos-vm/adapter and the (soon used to be parallel-)executor.

I am indeed happy now, even less logic spread around different crates. Tech debt --.

**Comment by gelash** [comment]
Split the PR, this is the first one that refactors the aptos-vm execution flow, removing unused status and merging sequential & parallel execution. Renamed parallel_executor -> block_executor as per @dariorussi 's suggestion, but since that's a lot of lines, made it a separate commit for the ease of reviewing.

@runtian-zhou can you have a look now?

**Comment by gelash** [comment]
@zekun000 @sasha8 ping ping

File: aptos-move/aptos-vm/src/block_executor/mod.rs
Comment (zekun000) [inline aptos-move/aptos-vm/src/block_executor/mod.rs]: why this is removed?

**Review by zekun000 — COMMENTED** [review]
lgtm, just some nits

File: aptos-move/aptos-vm/src/block_executor/mod.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/block_executor/mod.rs]: moved at the call-site in aptos_vm.rs, where we already have the logger, and now will log no matter how we execute:
https://github.com/aptos-labs/aptos-core/blob/82968ffdf6e772b7fa26eae01770662473773193/aptos-move/aptos-vm/src/aptos_vm.rs#L974

File: aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs
Comment (dariorussi) [inline aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs:32]: who is calling this function is a bit difficult to follow for me at the moment.
I think that must be around the problematic part you were mentioning offline...
I am still reviewing this, just fyi, and don't make it stop you from progressing...

File: aptos-move/block-executor/src/task.rs
Comment (dariorussi) [inline aptos-move/block-executor/src/task.rs:71]: this is annoying, you mentioned that and I am not clear what's going on, apology, I need a bit more time around this

**Review by dariorussi — COMMENTED** [review]
Some comments of appreciation :) and some real comments.
I am trying to understand the code and, of course, do not wait for me if you get enough approvals.
Back here soon.
Thanks so much for this, lots to like!

**Review by runtian-zhou — COMMENTED** [review]
Thanks for splitting the PR up! It seems to me a relatively clean change now and we it indeed seem to me a mare refactor and shouldn't be affecting any production logic.

File: aptos-move/aptos-vm/src/data_cache.rs
Comment (dariorussi) [inline aptos-move/aptos-vm/src/data_cache.rs:46]: not sure I follow the `Cow` story much but we'll talk about it

**Review by dariorussi — APPROVED** [review]
I spent a day on this, comparing with original, and I am not sure I see anything wrong, which is not to say there is not ;) 
I am going to accept this and @gelash we will talk about it today because I believe this should require a significant testing and verification.
But I personally do not want to hold you from being able to get in, though I would not want you to get in right now (my opinion of course).
In any case you need 2 reviews so my "approval" is just to make things simpler.
Point is, I do not believe I can add more in reviewing, considering the time I spent on this...

File: aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs
Comment (gelash) [reply aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs:475]: ^^

File: aptos-move/aptos-vm/src/aptos_vm.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/aptos_vm.rs:972]: uh, great catch, thanks a lot! I just verified and it was only reported in sequential mode and not in parallel (already problematic) so I guess I was trying to fix that, however, only in the success case, which I had missed the '?'. 

So let's keep the "on-success" behavior, but now it will work for both execution modes.

File: aptos-move/aptos-vm/src/block_executor/mod.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/block_executor/mod.rs:146]: RAYON_EXEC_POOL here is the same threadpool that we use in parallel execution. It has num_cpu threads but parallel execution uses concurrency_level many of them (by dispatching long-lived tasks).

It makes sense to use the same threadpool for other execution-related work, instead of going to the global rayon threadpool (which was happing before) as that could be less predictable and make performance unpredictable/unreliable.

File: aptos-move/aptos-vm/src/block_executor/mod.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/block_executor/mod.rs:169]: sequential execution can't return this error, it's the current hack (which will be oh so satisfying to get rid of) around the loader cache and parallel execution. Thus, if parallel returns this error it wasn't safe to execute the block in parallel, and sequential is supposed to clean it up. In fact, the match at the end of the function makes sure ModulePathReadWrite is unreachable.

File: aptos-move/aptos-vm/src/parallel_executor/mod.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/parallel_executor/mod.rs]: done

File: aptos-move/aptos-vm/src/parallel_executor/mod.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/parallel_executor/mod.rs]: it does seem like a better name!
many places in the aggregator crate where this is done, don't want to add more mess here, but added a TODO to one of the into()'s there and we can address later.

File: aptos-move/aptos-vm/src/data_cache.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/data_cache.rs:42]: We can get rid of the StateViewCache altogether: https://github.com/aptos-labs/aptos-core/commit/39f055092afc2b67afcd7990ba2bc6714b90682d#diff-c357acdc9a799c00c416262b0fd8e627e1f882f603e63d77625ae0f0fd9535fa

File: aptos-move/block-executor/src/task.rs
Comment (gelash) [reply aptos-move/block-executor/src/task.rs:71]: Indeed, need to create a common view on the executor side to deprecate both VersionedView and StateViewCache, and then we can unify the interfaces (and BTree or Multi-version becomes implementation detail). Also in the follow-up draft: https://github.com/aptos-labs/aptos-core/commit/39f055092afc2b67afcd7990ba2bc6714b90682d#diff-c357acdc9a799c00c416262b0fd8e627e1f882f603e63d77625ae0f0fd9535fa

File: aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs:32]: It is indeed the problematic part w. the separate *_btreemap_view and *_mvhashmap_view, but this we can unify in the follow-up (e.g. as in this draft https://github.com/aptos-labs/aptos-core/commit/39f055092afc2b67afcd7990ba2bc6714b90682d#diff-c357acdc9a799c00c416262b0fd8e627e1f882f603e63d77625ae0f0fd9535fa).

However, who calls execute_transaction is indeed very convoluted. I am adding a comment to explain.

File: aptos-move/aptos-vm/src/data_cache.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/data_cache.rs:58]: I intend to do this in a follow-up (get rid of StateViewCache and storage-wrapper/VersionedView altogether, merge inside block_executor and unify execute_transaction_* interfaces). Here is a draft commit for that follow-up: https://github.com/aptos-labs/aptos-core/commit/39f055092afc2b67afcd7990ba2bc6714b90682d

File: aptos-move/aptos-vm/src/data_cache.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/data_cache.rs:40]: done, and as mentioned, in the proposed-follow up (draft https://github.com/aptos-labs/aptos-core/commit/39f055092afc2b67afcd7990ba2bc6714b90682d), the whole StateViewCache bites the dust.

File: aptos-move/aptos-vm/src/block_executor/mod.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/block_executor/mod.rs:150]: I agree, I think this will be much more natural once I also hide the delta application and fallbacks inside the block_executor crate, then it can have execute_block() and internally deal with the concurrency_level (implementation detail). Now we kinda deal with some of it here, in the wrapper and some in the executor.

The challenges to getting there is simply to unify the output processing for both parallel and sequential modes (much like I am trying to unify the execute_transaction_* callback interfaces) and move it to block_executor. That would involve a bit of massaging of output_delta_resolver, which I can do as yet another separate diff.

**Comment by wrwg** [comment]
Lets give some more time for review from Move team members. (I know you filed this a while ago so sorry for no earlier attention.)

File: aptos-move/aptos-vm/src/block_executor/mod.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/block_executor/mod.rs:150]: here we go, draft follow-up to the draft follow-up I linked before:
https://github.com/aptos-labs/aptos-core/commit/d2e7f81f5c05ae4137104697e9cc964223c2a088
basically you can look how much simpler this file (aptos-vm/block_executor/mod.rs) becomes.

The logic moves to block_executor crate. This should make everything even nicer, and additionally can optimize the flow for generating which keys correspond to aggregators (previously required an exhaustive traversal over all transactions all affected keys, don't think it was even done in parallel), now we just filter & collect from MVHashMap keys and Delta resolution becomes an implementation detail.

**Comment by gelash** [comment]
> Lets give some more time for review from Move team members. (I know you filed this a while ago so sorry for no earlier attention.)

I was never going to land this without @runtian-zhou 's approval, but appreciate more eyes.

For context:
I already have 3 follow-ups implemented (2 drafts linked 2 in the comments, or can see here: https://github.com/aptos-labs/aptos-core/compare/main...seqparoutput, third for adding proptests which I separated from here for clarity) which intend to improve and simplify the whole executor <-> aptos-vm integration (+ sequential now), delta resolution, etc. 

But it all starts here, and once we have all executor flow nicely in the block_executor, opens up all the extension and gas hooks that we need for different outstanding tasks (can explain offline).

File: aptos-move/aptos-vm/src/parallel_executor/mod.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/parallel_executor/mod.rs]: Will optimize this flow completely soon (in 2 PRs time):
https://github.com/aptos-labs/aptos-core/commit/b423013ebba88a6362add5aa0a6077fe01edcb41

File: aptos-move/aptos-vm/src/block_executor/mod.rs
Comment (wrwg) [inline aptos-move/aptos-vm/src/block_executor/mod.rs:39]: Not a big fan of avoiding the orphan rule which is there for a reason (and not a technical one). It makes perhaps notations a bit more convenient, but generally the code harder to understand. Not saying this needs to be differently done, but just an opinion.

File: aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs
Comment (wrwg) [inline aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs:29]: This `S` is always a `StateView`, right? Then this is really strange and the need for this wrapper type just demonstrates how broken the AptosVM is (and requires a larger overhaul). Because if you look at `AptosVM`, you see that it wraps `AptosVMImpl`, which in turn is created from a state view. That state view is then hidden inside of the storage adapter.  The need to be able to get hand a StateView already created acrobatic code in other places.

This is probably another set of PRs after this one, but I really wish we could drastically simplify the architecture here:

- Only one `AptosVM` (no VMImpl, BlockVM, MoveVmExt, and other complications)
- That `AptosVM` implements all the parallel and sequential execution logic. Multiple `impl AptosVM` split over files can help to tam the complexity.

Because parallel execution is build into our Move adapter there is really no need to maintain a code layer underneath without PE.

File: aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs
Comment (wrwg) [inline aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs:76]: If you have the `AptosVMWrapper` to be actually mostly an executortask, perhaps call it like this (`struct AptosExecutorTask`?)

**Review by wrwg — COMMENTED** [review]
This generally seems to go into the right direction (that is unifying sequential/parallel block execution). We need a larger refactoring of the adapter (see comment below), but this PR can be an incremental step for this. Leaving detail review to folks more acquainted with PE.

**Comment by gelash** [comment]
> This generally seems to go into the right direction (that is unifying sequential/parallel block execution). We need a larger refactoring of the adapter (see comment below), but this PR can be an incremental step for this. Leaving detail review to folks more acquainted with PE.

Totally agreed, that's precisely the intention. There are some specific fixes, but regarding the overall refactoring aspect, my rationale at the moment is to try and make some incremental local simplifications / improvements while hopefully going in the good global direction (allow having a single "executor" that can be cleanly connected to other parts of aptos-vm, re-used across blocks, etc). Hopefully with some iteration (and there are indeed 2/3 PRs coming right after), we will also start seeing the big picture better. One heuristic for me for now is to push some pieces of logic to the block_executor side (e.g. in the next PR, the StateView wrapper business currently done in different ways in aptos-vm), since that code is newer and should have more fixed structure / flow. 

However, then we should absolutely look into precisely the question of how the block executor is incorporated into aptos_vm. For some context, I believe the current state of affairs is due to two things - @runtian-zhou can confirm or deny as the author and the ultimate expert on the wrapper flow:
(a) make parallel executor standalone crate with generic parameters - be able to test without Move-VM.
(b) a clear separator of abstractions to facilitate the development at the time.
The testing (and prop-testing) framework is probably the best thing that we got out of building it this way. But now all these layers in aptos-vm / aptos_vm_impl / adapter / wrapper need to eventually also be simplified, especially since we have a lot of use-cases where we'd need hooks to & from other parts of aptos-vm.

File: aptos-move/aptos-vm/src/block_executor/mod.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/block_executor/mod.rs:39]: I really dislike this as well and tried to get rid of it, can solve some issues but the problem I couldn't easily overcome is a potential circular dependency between aptos-aggregator (where TransactionOutputExt is defined) and block-executor crates (that defines traits, and also uses aggregators).. Hopefully we will get to a place where we restructure this wrapper, or in general restructure the crates and factor out the common types (like we do currently at the aptos-core level, something similar maybe we could do the same at the aptos-vm or aptos-move).

File: aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs
Comment (gelash) [reply aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs:76]: makes sense! done.

**Comment by runtian-zhou** [comment]
The way how parallel executor is structured like this is exactly what @gelash suggests. The whole intention of this `parallel_executor` crate abstraction is to:
1. Help testing core scheduling logic without worrying about the Aptos VM implementation.
2. Being able to run benchmarks that are independent of the AptosVM.

The testing aspect is probably more important in this context cuz it's generally quite hard to test parallel code like this.

File: aptos-move/block-executor/src/errors.rs
Comment (runtian-zhou) [inline aptos-move/block-executor/src/errors.rs:8]: Why is this error removed?

File: aptos-move/block-executor/src/executor.rs
Comment (runtian-zhou) [inline aptos-move/block-executor/src/executor.rs:362]: nit: `&[T]`

File: aptos-move/block-executor/src/executor.rs
Comment (runtian-zhou) [inline aptos-move/block-executor/src/executor.rs:370]: Was wondering if those assertion would fail in production as they are a good source for the network availability attack. Would it be better to return an error when such condition is violated?

**Review by runtian-zhou — COMMENTED** [review]
Almost looks good to me! Agreed with @wrwg  that we need more refactoring here but this is a great step moving forward!

File: aptos-move/block-executor/src/executor.rs
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:362]: done

File: aptos-move/block-executor/src/errors.rs
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:8]: Was never used - happy to bring it back whenever needed.

File: aptos-move/block-executor/src/executor.rs
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:370]: Currently, must be true due to 
https://github.com/aptos-labs/aptos-core/blob/e197a64f990b349b888eec4624c1adf945d0ef67/aptos-move/aptos-vm/src/block_executor/mod.rs#L150, and second follow-up PR moves this dispatching inside block_executor crate: https://github.com/aptos-labs/aptos-core/blob/b423013ebba88a6362add5aa0a6077fe01edcb41/aptos-move/block-executor/src/executor.rs#L377, and execute_parallel becomes pub(crate), so we could also just delete the assert at that point. In fact, I like that idea to delete just this assert then.

If you really think we should return an error here, let me know what kind of error and what should happen in that case, we skip the whole block? But it will complicate some places and we should worry about determinism and probably not worth for some invariant that trivially should never happen.

**Review by zekun000 — APPROVED** [review]

**Pull request merged**
